xquant/keras/keras_report_utils.py

```diff
@@ -0,0 +1,380 @@
+            """
+            Extract activations from specified layers of the model for the given input data.
+
+            Args:
+                model (Model): The Keras model from which to extract activations.
+                layer_names (List[str]): List of layer names for which activations are to be extracted.
+                data (Any): Input data for which activations are to be computed.
+
+            Returns:
+                Dict[str, np.ndarray]: A dictionary mapping layer names to their corresponding activations.
+            """
+            # Create a new model that outputs the activations of the specified layers
+            intermediate_layer_model = Model(inputs=model.input,
+                                             outputs=[model.get_layer(name).output for name in layer_names])
+
+            predictions = intermediate_layer_model.predict(data)
+
+            # Map the activations to their corresponding layer names
+            return {layer_name: predictions[i] for i, layer_name in enumerate(layer_names)}
+
+        metrics_to_compute = list(self.get_default_metrics().keys())
+        if custom_metrics_intermediate:
+            if not isinstance(custom_metrics_intermediate, dict):
+                Logger.critical(f"custom_metrics_intermediate should be a dictionary but is {type(custom_metrics_intermediate)}")
+
+            metrics_to_compute += list(custom_metrics_intermediate.keys())
+
+        float_name2quant_name = self.get_float_to_quantized_compare_points(float_model=float_model,
+                                                                           quantized_model=quantized_model)
+
+        results = {q_layer: [] for q_layer in float_name2quant_name.values()}
+
+        for x in dataset():
+            quant_activations = get_activations(quantized_model, list(float_name2quant_name.values()), x)
+            float_activations = get_activations(float_model, list(float_name2quant_name.keys()), x)
+
+            for float_layer, quant_layer in float_name2quant_name.items():
+                float_activation = float_activations[float_layer]
+                quant_activation = quant_activations[quant_layer]
+
+                results[quant_layer].append(
+                    self.compute_metrics((float_activation, quant_activation), custom_metrics_intermediate))
+
+        aggregated_metrics = {}
+        for layer_name, layer_metrics in results.items():
+            combined_dict = {}
+            for item in layer_metrics:
+                for key, value in item.items():
+                    if key not in combined_dict:
+                        combined_dict[key] = []
+                    combined_dict[key].append(value)
+            for k, v in combined_dict.items():
+                combined_dict[k] = np.mean(v)
+            aggregated_metrics[layer_name] = combined_dict
+
+        return aggregated_metrics
+
+    def get_default_metrics(self) -> Dict[str, Callable]:
```

Comment:
This is a good example for why I think that this class needs to be split into modules - 
it is very reasonable that we will need to use this similarity functions in the future in other places in the project, I don't think that they should be "hidden" as nested functions inside this class.

Extract them to a "similarity_metrics" file, that way it would be simpler to use them, maintain and extend them.